Descent3/init.cpp

```diff
@@ -1413,33 +1424,63 @@ static const tTempFileInfo temp_file_wildcards[] = {{"d3s*.tmp"}, {"d3m*.tmp"},
                                                     {"d3c*.tmp"}, {"d3t*.tmp"}, {"d3i*.tmp"}};
 static const int num_temp_file_wildcards = sizeof(temp_file_wildcards) / sizeof(tTempFileInfo);
 
+
+// Get the binary path
+char exec_path[_MAX_PATH];
+char* getExecutablePath(char *exec_path, size_t len) {
+#ifdef _WIN32
+    if (GetModuleFileName(NULL, exec_path, _MAX_PATH) == 0) {
+        DWORD error = GetLastError();
+        Error("GetModuleFileName failed!");
+        return NULL;
+    }
+    exec_path[_MAX_PATH - 1] = '\0';
+    return exec_path;
+#else
+    if (realpath("/proc/self/exe", exec_path) == NULL) {
+        perror("realpath");
+        return NULL;
+    }
+    exec_path[_MAX_PATH - 1] = '\0';
+    return exec_path;
+#endif
+}
+
 /*
         I/O systems initialization
 */
 void InitIOSystems(bool editor) {
   ddio_init_info io_info;
-  int dirlen = _MAX_PATH;
-
-  // Set the base directory
-  int dirarg = FindArg("-setdir");
-  int exedirarg = FindArg("-useexedir");
-  if (dirarg) {
-    strcpy(Base_directory, GameArgs[dirarg + 1]);
-  } else if (exedirarg) {
-    strcpy(Base_directory, GameArgs[0]);
-    int len = strlen(Base_directory);
-    for (int i = (len - 1); i >= 0; i--) {
-      if (i == '\\') {
-        Base_directory[i] = '\0';
-      }
+
+    // Set the base directory
+    int dirarg = FindArg("-setdir");
+    int exedirarg = FindArg("-useexedir");
+    if (dirarg) {
+        strncpy(Base_directory, GameArgs[dirarg + 1], sizeof(Base_directory) - 1);
+        Base_directory[sizeof(Base_directory) - 1] = '\0';
+    } else if (exedirarg) {
+        memset(exec_path, 0, sizeof(exec_path));
+        // Populate exec_path with the executable path
+        if (getExecutablePath(exec_path, sizeof(exec_path)) == NULL) {
+            Error("Failed to get executable path\n");
+        } else {
+            // Extract directory from the executable path
```

Comment:
Probably we can use here `std::filesystem` functions from C++17, code would much cleaner and easier.
Like

```cpp
Base_directory = std::filesystem::path::root_path(Base_directory).c_str();
```

Reply:
It's not really any less lines, but it works.